File: macos/Sources/ChoroReader/Search/SearchIndex.swift
import Foundation

/// 全文検索の索引。
///
/// 日本語には単語の切れ目が無いので、二字組（バイグラム）を鍵にした転置索引を持つ。
///
/// この索引は**候補を絞るだけ**で、当たりを決めない。
/// 二字組が同じ単位に出ていても、続けて出ているとは限らないためである。
/// 絞った単位を `DocumentSearch` の走査にかけ直して初めて当たりが決まる。
/// そうすることで、結果は索引が無いときと一字も変わらない。
///
/// 当たりの位置を索引に持たせないぶん、容量が小さい。
/// 日本語 50 万字（400 ページ相当）で 300 KB 前後に収まる。
struct SearchIndex {
    /// 単位の終わりを表す番人。
    ///
    /// 1 文字だけの問い合わせは「その字で始まる二字組」をまとめて引くが、
    /// 単位の最後の 1 文字は次の字を持たないので、そのままでは引けない。
    /// 二字目を 0 にした組を足しておくと、範囲を引くだけで漏れなく当たる。
    private static let sentinel: UInt32 = 0

    private(set) var unitCount: Int
    private var keys: [UInt64]
    private var offsets: [UInt32]
    private var units: [UInt32]

    var isEmpty: Bool { keys.isEmpty }

    // MARK: - 組み立て

    static func build(_ unitTexts: [String]) -> SearchIndex {
        var pairs: [(key: UInt64, unit: UInt32)] = []

        for (number, text) in unitTexts.enumerated() {
            let number = UInt32(number)
            let scalars = SearchFold.scalars(text)
            guard !scalars.isEmpty else { continue }
            for i in 0 ..< (scalars.count - 1) {
                pairs.append((key(scalars[i], scalars[i + 1]), number))
            }
            pairs.append((key(scalars[scalars.count - 1], sentinel), number))
        }

        pairs.sort { $0.key == $1.key ? $0.unit < $1.unit : $0.key < $1.key }

        var index = SearchIndex(unitCount: unitTexts.count, keys: [], offsets: [0], units: [])
        for pair in pairs {
            if index.keys.last == pair.key {
                if index.units.last == pair.unit { continue }
            } else {
                index.keys.append(pair.key)
                index.offsets.append(UInt32(index.units.count))
            }
            index.units.append(pair.unit)
            index.offsets[index.offsets.count - 1] = UInt32(index.units.count)
        }
        return index
    }

    // MARK: - 引き

    /// 問い合わせが当たりうる単位の番号。昇順。
    ///
    /// `nil` は「索引では絞れないので全部を見よ」を意味する。
    func candidates(_ query: String) -> [Int]? {
        let scalars = SearchFold.scalars(query)
        if scalars.isEmpty { return nil }
        if scalars.count == 1 { return byFirstScalar(scalars[0]) }

        var found: [UInt32]?
        for i in 0 ..< (scalars.count - 1) {
            let posting = self.posting(Self.key(scalars[i], scalars[i + 1]))
            found = found.map { Self.intersect($0, posting) } ?? Array(posting)
            if found?.isEmpty == true { break }
        }
        return found?.map(Int.init)
    }

    private func posting(_ wanted: UInt64) -> ArraySlice<UInt32> {
        var low = 0
        var high = keys.count
        while low < high {
            let middle = (low + high) / 2
            if keys[middle] < wanted { low = middle + 1 } else { high = middle }
        }
        guard low < keys.count, keys[low] == wanted else { return [] }
        return units[Int(offsets[low]) ..< Int(offsets[low + 1])]
    }

    private func byFirstScalar(_ scalar: UInt32) -> [Int] {
        let from = UInt64(scalar) << 21
        let to = from + (1 << 21)
        var seen = [Bool](repeating: false, count: unitCount)
        var found: [Int] = []
        for at in lowerBound(from) ..< lowerBound(to) {
            for unit in units[Int(offsets[at]) ..< Int(offsets[at + 1])] where !seen[Int(unit)] {
                seen[Int(unit)] = true
                found.append(Int(unit))
            }
        }
        return found.sorted()
    }

    private func lowerBound(_ wanted: UInt64) -> Int {
        var low = 0
        var high = keys.count
        while low < high {
            let middle = (low + high) / 2
            if keys[middle] < wanted { low = middle + 1 } else { high = middle }
        }
        return low
    }

    private static func intersect(_ left: [UInt32], _ right: ArraySlice<UInt32>) -> [UInt32] {
        var found: [UInt32] = []
        var a = left.startIndex
        var b = right.startIndex
        while a < left.endIndex, b < right.endIndex {
            if left[a] < right[b] {
                a += 1
            } else if left[a] > right[b] {
                b += 1
            } else {
                found.append(left[a])
                a += 1
                b += 1
            }
        }
        return found
    }

    private static func key(_ first: UInt32, _ second: UInt32) -> UInt64 {
        (UInt64(first) << 21) | UInt64(second)
    }

    // MARK: - 書き出しと読み込み
    //
    // 鍵も単位番号も昇順に並んでいるので、差分を可変長で書けば大半が 1 バイトで済む。
    // 形式は Tauri 版と同じにしてある。ファイルを共有はしないが、決めを 2 度書かないため。

    private static let magic: [UInt8] = Array("CHIX".utf8)
    private static let version: UInt8 = 1

    func encoded() -> Data {
        var out = Data(Self.magic)
        out.append(Self.version)
        put(&out, UInt64(unitCount))
        put(&out, UInt64(keys.count))

        var previousKey: UInt64 = 0
        for at in keys.indices {
            put(&out, keys[at] - previousKey)
            previousKey = keys[at]
            put(&out, UInt64(offsets[at + 1] - offsets[at]))
        }

        for at in keys.indices {
            var previousUnit: UInt32 = 0
            for unit in units[Int(offsets[at]) ..< Int(offsets[at + 1])] {
                put(&out, UInt64(unit - previousUnit))
                previousUnit = unit
            }
        }
        return out
    }

    init?(decoding data: Data) {
        let bytes = [UInt8](data)
        guard bytes.count > Self.magic.count,
              Array(bytes[0 ..< Self.magic.count]) == Self.magic,
              bytes[Self.magic.count] == Self.version else { return nil }

        var cursor = Self.magic.count + 1
        func get() -> UInt64? {
            var value: UInt64 = 0
            var shift: UInt64 = 0
            while cursor < bytes.count {
                let byte = bytes[cursor]
                cursor += 1
                value |= UInt64(byte & 0x7f) << shift
                if byte & 0x80 == 0 { return value }
                shift += 7
                if shift >= 64 { return nil }
            }
            return nil
        }

        guard let count = get(), let keyCount = get(), count <= UInt64(UInt32.max) else { return nil }
        unitCount = Int(count)

        keys = []
        keys.reserveCapacity(Int(keyCount))
        var lengths: [UInt32] = []
        lengths.reserveCapacity(Int(keyCount))
        var previousKey: UInt64 = 0
        for _ in 0 ..< keyCount {
            guard let delta = get(), let length = get() else { return nil }
            previousKey += delta
            keys.append(previousKey)
            lengths.append(UInt32(truncatingIfNeeded: length))
        }

        offsets = [0]
        offsets.reserveCapacity(Int(keyCount) + 1)
        units = []
        for length in lengths {
            var previousUnit: UInt32 = 0
            for _ in 0 ..< length {
                guard let delta = get() else { return nil }
                previousUnit += UInt32(truncatingIfNeeded: delta)
                guard previousUnit < UInt32(unitCount) else { return nil }
                units.append(previousUnit)
            }
            offsets.append(UInt32(units.count))
        }
    }

    private init(unitCount: Int, keys: [UInt64], offsets: [UInt32], units: [UInt32]) {
        self.unitCount = unitCount
        self.keys = keys
        self.offsets = offsets
        self.units = units
    }

    private func put(_ out: inout Data, _ value: UInt64) {
        var value = value
        while true {
            let byte = UInt8(value & 0x7f)
            value >>= 7
            if value == 0 {
                out.append(byte)
                return
            }
            out.append(byte | 0x80)
        }
    }
}

/// 索引に載せる形。走査側と同じ規則で畳む。
///
/// 走査は `String.range(of:options:)` に任せているので、畳み方もその `options` に合わせる。
/// ここが走査より厳しくなると、索引が本当の当たりを落としてしまう。
enum SearchFold {
    /// 合成の違いを消してから 1 文字ずつ数に落とす。
    ///
    /// 正規化しておけば、「か」＋濁点と「が」が同じ数になる。
    /// 結合文字が後ろに付く文字は先頭の符号位置だけを見るが、
    /// それで一致する組が増える方向にしか動かないので、当たりを落とすことはない。
    static func scalars(_ text: String) -> [UInt32] {
        text.folding(options: DocumentSearch.options, locale: nil)
            .precomposedStringWithCanonicalMapping
            .compactMap { $0.unicodeScalars.first?.value }
    }
}

File: windows/ChoroReader.Core/SearchIndex.cs
using System.Text;

namespace ChoroReader.Core;

/// <summary>
/// 全文検索の索引。
///
/// <para>
/// 日本語には単語の切れ目が無いので、二字組（バイグラム）を鍵にした転置索引を持つ。
/// </para>
/// <para>
/// この索引は<b>候補を絞るだけ</b>で、当たりを決めない。
/// 二字組が同じ単位に出ていても、続けて出ているとは限らないためである。
/// 絞った単位を <see cref="DocumentSearch"/> の走査にかけ直して初めて当たりが決まる。
/// そうすることで、結果は索引が無いときと一字も変わらない
/// （conformance/CONTRACT.md の「検索のヒット位置、件数、順序」を索引の導入で動かさないため）。
/// </para>
/// <para>
/// 当たりの位置を索引に持たせないぶん、容量が小さい。
/// 索引の形式は実装ごとに閉じたもので、契約には載せない（spec.md 10.4）。
/// 揃えるべきは索引の中身ではなく検索結果である。
/// </para>
/// </summary>
public sealed class SearchIndex
{
    /// <summary>索引の形式の版。読めない版のファイルは捨てて作り直す。</summary>
    private const byte Version = 1;

    private static ReadOnlySpan<byte> Magic => "CHIX"u8;

    /// <summary>
    /// 単位の終わりを表す番人。
    ///
    /// 1 文字だけの問い合わせは「その字で始まる二字組」をまとめて引くが、
    /// 単位の最後の 1 文字は次の字を持たないので、そのままでは引けない。
    /// 二字目を 0 にした組を足しておくと、範囲を引くだけで漏れなく当たる。
    /// </summary>
    private const uint Sentinel = 0;

    private readonly uint _unitCount;

    /// <summary>昇順に並べた鍵。</summary>
    private readonly ulong[] _keys;

    /// <summary><c>_keys[i]</c> の単位番号は <c>_units[_offsets[i].._offsets[i + 1]]</c>。</summary>
    private readonly int[] _offsets;

    private readonly uint[] _units;

    private SearchIndex(uint unitCount, ulong[] keys, int[] offsets, uint[] units)
    {
        _unitCount = unitCount;
        _keys = keys;
        _offsets = offsets;
        _units = units;
    }

    public int UnitCount => (int)_unitCount;

    public bool IsEmpty => _keys.Length == 0;

    /// <summary>2 文字を 1 つの数にする。Unicode の符号位置は 21 bit に収まる。</summary>
    private static ulong KeyOf(Rune first, uint second) => ((ulong)first.Value << 21) | second;

    public static SearchIndex Build(IReadOnlyList<string> units)
    {
        var pairs = new List<(ulong Key, uint Unit)>();

        for (var number = 0; number < units.Count; number++)
        {
            var chars = Fold.Text(units[number]);
            for (var i = 0; i + 1 < chars.Count; i++)
            {
                pairs.Add((KeyOf(chars[i], (uint)chars[i + 1].Value), (uint)number));
            }
            if (chars.Count > 0)
            {
                pairs.Add((KeyOf(chars[^1], Sentinel), (uint)number));
            }
        }

        pairs.Sort();

        var keys = new List<ulong>();
        var offsets = new List<int> { 0 };
        var postings = new List<uint>();
        (ulong Key, uint Unit)? previous = null;

        foreach (var pair in pairs)
        {
            if (previous == pair)
            {
                continue;
            }
            previous = pair;

            if (keys.Count == 0 || keys[^1] != pair.Key)
            {
                keys.Add(pair.Key);
                offsets.Add(postings.Count);
            }
            postings.Add(pair.Unit);
            offsets[^1] = postings.Count;
        }

        return new SearchIndex((uint)units.Count, [.. keys], [.. offsets], [.. postings]);
    }

    /// <summary>
    /// 問い合わせが当たりうる単位の番号。
    ///
    /// <para>null は「索引では絞れないので全部を見よ」を意味する。</para>
    /// </summary>
    public IReadOnlySet<int>? Candidates(string query)
    {
        var chars = Fold.Text(query);
        if (chars.Count == 0)
        {
            return null;
        }
        if (chars.Count == 1)
        {
            return ToSet(ByFirstChar(chars[0]));
        }

        uint[]? found = null;
        for (var i = 0; i + 1 < chars.Count; i++)
        {
            var posting = Posting(KeyOf(chars[i], (uint)chars[i + 1].Value));
            found = found is null ? posting : Intersect(found, posting);
            if (found.Length == 0)
            {
                break;
            }
        }
        return found is null ? null : ToSet(found);
    }

    private static IReadOnlySet<int> ToSet(uint[] units)
    {
        var set = new HashSet<int>(units.Length);
        foreach (var unit in units)
        {
            set.Add((int)unit);
        }
        return set;
    }

    private uint[] Posting(ulong key)
    {
        var at = Array.BinarySearch(_keys, key);
        return at < 0 ? [] : _units[_offsets[at].._offsets[at + 1]];
    }

    /// <summary>その字で始まる二字組をすべて合併する。</summary>
    private uint[] ByFirstChar(Rune c)
    {
        var from = (ulong)c.Value << 21;
        var to = from + (1UL << 21);
        var start = LowerBound(from);
        var end = LowerBound(to);

        var seen = new bool[_unitCount];
        var found = new List<uint>();
        for (var at = start; at < end; at++)
        {
            for (var i = _offsets[at]; i < _offsets[at + 1]; i++)
            {
                var unit = _units[i];
                if (!seen[unit])
                {
                    seen[unit] = true;
                    found.Add(unit);
                }
            }
        }
        found.Sort();
        return [.. found];
    }

    /// <summary><paramref name="value"/> 以上が最初に現れる位置。</summary>
    private int LowerBound(ulong value)
    {
        var low = 0;
        var high = _keys.Length;
        while (low < high)
        {
            var middle = low + ((high - low) / 2);
            if (_keys[middle] < value)
            {
                low = middle + 1;
            }
            else
            {
                high = middle;
            }
        }
        return low;
    }

    private static uint[] Intersect(uint[] left, uint[] right)
    {
        var found = new List<uint>();
        int a = 0, b = 0;
        while (a < left.Length && b < right.Length)
        {
            if (left[a] < right[b])
            {
                a++;
            }
            else if (left[a] > right[b])
            {
                b++;
            }
            else
            {
                found.Add(left[a]);
                a++;
                b++;
            }
        }
        return [.. found];
    }

    // MARK: 書き出しと読み込み
    //
    // 鍵も単位番号も昇順に並んでいるので、差分を可変長で書けば大半が 1 バイトで済む。

    public byte[] Encode()
    {
        var output = new List<byte>();
        output.AddRange(Magic);
        output.Add(Version);
        Varint.Put(output, _unitCount);
        Varint.Put(output, (ulong)_keys.Length);

        var previous = 0ul;
        for (var at = 0; at < _keys.Length; at++)
        {
            Varint.Put(output, _keys[at] - previous);
            previous = _keys[at];
            Varint.Put(output, (ulong)(_offsets[at + 1] - _offsets[at]));
        }

        for (var at = 0; at < _keys.Length; at++)
        {
            var last = 0u;
            for (var i = _offsets[at]; i < _offsets[at + 1]; i++)
            {
                Varint.Put(output, _units[i] - last);
                last = _units[i];
            }
        }
        return [.. output];
    }

    /// <summary>読めなければ null。壊れたファイルは捨てて作り直す。</summary>
    public static SearchIndex? Decode(byte[] bytes)
    {
        var head = Magic.Length + 1;
        if (bytes.Length < head || !bytes.AsSpan(0, Magic.Length).SequenceEqual(Magic)
            || bytes[Magic.Length] != Version)
        {
            return null;
        }

        var cursor = new Varint.Cursor(bytes, head);
        if (cursor.Get() is not { } rawUnitCount || rawUnitCount > uint.MaxValue
            || cursor.Get() is not { } rawKeyCount || rawKeyCount > int.MaxValue)
        {
            return null;
        }
        var unitCount = (uint)rawUnitCount;
        var keyCount = (int)rawKeyCount;

        var keys = new ulong[keyCount];
        var lengths = new int[keyCount];
        var previous = 0ul;
        for (var i = 0; i < keyCount; i++)
        {
            if (cursor.Get() is not { } step || cursor.Get() is not { } length || length > int.MaxValue)
            {
                return null;
            }
            previous += step;
            keys[i] = previous;
            lengths[i] = (int)length;
        }

        var offsets = new int[keyCount + 1];
        var units = new List<uint>();
        for (var i = 0; i < keyCount; i++)
        {
            var last = 0u;
            for (var n = 0; n < lengths[i]; n++)
            {
                if (cursor.Get() is not { } step || step > uint.MaxValue)
                {
                    return null;
                }
                last += (uint)step;
                if (last >= unitCount)
                {
                    return null;
                }
                units.Add(last);
            }
            offsets[i + 1] = units.Count;
        }

        return new SearchIndex(unitCount, keys, offsets, [.. units]);
    }
}
